packages/async-rewriter2/src/async-writer-babel.ts

```
@@ -0,0 +1,575 @@
+  // of helpers which are available inside the function.
+  const identifierGroupKey = '@@mongosh.identifierGroup';
+
+  // We fetch the symbol constructor as
+  //   Object.getOwnPropertySymbols(Array.prototype)[0].constructor
+  // because Symbol refers to BSONSymbol inside the target mongosh context.
+  // (This is the only mongosh-specific hack in here.)
+  const symbolConstructor = t.memberExpression(t.memberExpression(
+    t.callExpression(
+      t.memberExpression(t.identifier('Object'), t.identifier('getOwnPropertySymbols')),
+      [t.memberExpression(t.identifier('Array'), t.identifier('prototype'))]
+    ),
+    t.identifier('0'),
+    true),
+  t.identifier('constructor'));
+
+  return {
+    visitor: {
+      BlockStatement(path) {
+        // This might be a function body. If it's what we're looking for, wrap it.
+        if (!path.parentPath.isFunction()) return;
+        // Don't wrap things we've already wrapped.
+        if (path.parentPath.getData(identifierGroupKey)) return;
+        // Don't wrap the inner function we've created while wrapping another function.
+        if (path.parentPath.node[isGeneratedInnerFunction]) return;
+        // Don't wrap helper functions with async-rewriter-generated code.
+        if (path.parentPath.node[isGeneratedHelper]) return;
+        // Don't wrap generator functions. There is no good way to handle them.
+        if (path.parentPath.node.generator && !path.parentPath.node.async) return;
+        // Finally, do not wrap constructor functions. This is not a technical
+        // necessity, but rather a result of the fact that we can't handle
+        // asynchronicity in constructors well (e.g.: What happens when you
+        // subclass a class with a constructor that returns asynchronously?).
+        if (path.parentPath.isClassMethod() &&
+            path.parentPath.node.key.type === 'Identifier' &&
+            path.parentPath.node.key.name === 'constructor') {
+          return;
+        }
```

These are cases that we may want to handle by also checking any expression for being “synthetic” Promises, but where we’d probably want to throw an exception if turns out to be one.

---

Opened https://jira.mongodb.org/browse/MONGOSH-654 as a sub-task to track this